pkg/storage/mock_helpers_test.go

```diff
@@ -38,3 +39,13 @@ func inOrder(calls ...interface{}) mockSequence {
 	}
 	return mockSequence{first, last}
 }
+
+// containers/image/storage.storageReference.StringWithinTransport
+func mockStorageReferenceStringWithinTransport(storeMock *containerstoragemock.MockStore) mockSequence {
+	return inOrder(
+		storeMock.EXPECT().GraphOptions().Return([]string{}),
```

Comment:
This is just an idea and not something we need to do right now:

Do you think it would be worth adding a mock'able public API interface to containers/{image,storage}? I think we could then avoid these deeply nested mocks a bit.

Reply:
Mockable in what way?  (See the disclaimer elsewhere: I don’t quite know what I am doing at this conceptual level.)

c/image does have a set of `types.Image{Transport,Reference,Source,Destination}` interfaces, and (although that’s a bit unusual for c/image), the `storage.Transport` even allows replacing the underlying `c/storage.Store` object), so it should be quite possible to write tests that ensure that c/image is invoked correctly (and that exercise the `c/image/copy.Image` function in full, perhaps), without actually having to execute the full implementation of the `c/image/storage.storageTransport` object: basically, somewhere in CRI-O, parametrize whatever creates `types.ImageReference` objects to return an `ImageReference` object that points to a test-only mock transport purpose-built for the test, instead of a real one.  For the most part, that would mean wrapping `alltransports.ParseImageName` (and probably `c/image/storage.Transport.ParseStoreReference`, and maybe a few others) in an interface that would allow CRI-O tests to return a mock instead of the real thing.

AFAICS this would be something that CRI-O needs to do; as long as CRI-O calls the `alltransports.ParseImageName` global directly, the return value of that can’t be mocked (well, without modifying `ParseImageName` to refer to global state, at least, which would be fairly ugly).

Reply:
> Mockable in what way? (See the disclaimer elsewhere: I don’t quite know what I am doing at this conceptual level.)

In an ideal world, which we'll probably never have, we would unit-test only the code part which contains to a dedicated function. Everything else could be mocked, like IO side effects. The issue is now that we can only mock public interfaces in golang, which is not available in every case (like for reading and writing of files). This means that we have to go down the path which is out of scope of the function and this makes it hard to understand/write. In terms of unit testing, it would be great to consume an API which provides an public mockable interface at its outermost endpoints.